instance/models/mixins/redis.py

```diff
@@ -0,0 +1,158 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015-2021 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app model mixins - Redis
+"""
+
+# Imports #####################################################################
+
+from functools import lru_cache
+import redis
+import json
+import string
+import urllib.parse
+
+from django.db import IntegrityError, models
+from django.utils.crypto import get_random_string
+import requests
+
+from instance.models.redis import RedisUser
+from instance.models.redis_server import RedisServer
+
+
+# Functions ###################################################################
+
+def new_redis_user() -> RedisUser:
+    """
+    Return a new Redis user.
+    """
+    return RedisUser.objects.create(
+        username=get_random_string(length=32, allowed_chars=string.ascii_lowercase),
+        password=get_random_string(length=64),
+    )
+
+
+def select_random_redis_server():
+    """
+    Helper for the field default of `redis_server`.
+    """
+    return RedisServer.objects.select_random().pk
+
+
+# Classes #####################################################################
+
+class RedisAPIError(Exception):
+    """Exception indicating that a call to the Redis API failed."""
+
+
+class RedisInstanceMixin(models.Model):
+    """
+    An instance that uses a Redis vhost with a set of users.
+    """
+    redis_server = models.ForeignKey(
+        RedisServer,
+        null=True,
+        blank=True,
+        default=select_random_redis_server,
```

Comment:
@gabor-boros oh _this_ is where the select random server is used.  Would we be able to rework that so it selects the redis server defined by `DEFAULT_INSTANCE_REDIS_URL` by default?

Reply:
@swalladge That would affect all the classes that are inherited from `SharedServerManager` (`LoadBalancingServerManager`, `RabbitMQServerManager`, `RedisServerManager` and `DatabaseServerManager`). I'm happy to rework, but I want to double-check: Are you sure we want to refactor that logic as part of this PR?

Reply:
Yeah ok, good idea to leave it for now then :+1: